### DISCUSSION TITLE
xct:TouchEffect.LongPressCommand on CollectionView not working on CollectionItem?

### DISCUSSION THREAD

**Discussion by klingu** [xamarin/XamarinCommunityToolkit#1554] [Q&A] [answered]
@AndreiMisiukevich 

Hi Andrei,

I've watched your video with Gerald from June 8th. Great example and great feature to donate to the Xamarin Community Toolkit!

Can you please help me out on implementing the LongPressCommand on the CollectionView.
I've tried on many levels in the CollectionView and the DataTemplate. But It will not fire the command.
I also have Googled for a proper solution but it did not show up...

On a Label on the same page it does fire the command, so I know it works from there.

Xaml code
```
<ContentPage.Content>
        <StackLayout Spacing="0" BackgroundColor="{DynamicResource PageBgColor}">
            <Label Text="PRESSING THIS LABEL FOR 1 SECOND FIRES THE COMMAND"
                          xct:TouchEffect.LongPressCommand="{Binding LongPressCommand}"
                          xct:TouchEffect.LongPressCommandParameter="Label"
                          xct:TouchEffect.LongPressDuration="1000" />
            <CollectionView x:Name="MessagesListViewGrouped">
                <CollectionView.ItemTemplate >
                    <DataTemplate x:DataType="models:Message">
                        <viewcells:MessageListViewCell/>
                    </DataTemplate>
                </CollectionView.ItemTemplate>
            </CollectionView>

```

Code ViewModel
```
        public ICommand LongPressCommand => new Command((s) =>
        {
            if (Device.Android == Device.RuntimePlatform)
            {
                Debug.WriteLine($">>> LONG PRESS COMMAND FROM {(string)s}");
            }
        });
```

MessageListViewCell Xaml Code is based on a Grid and has data triggers on it. and a StackLayout
```
<Grid xmlns="http://xamarin.com/schemas/2014/forms"
      xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
      x:Class="ViewLayer.ViewCells.MessageListViewCell"
      x:Name="ViewCell"
      xmlns:effects="clr-namespace:MO.Shared.Views.Effects">
    <Grid.Triggers>
        <DataTrigger Binding="{Binding IsSelected}" TargetType="Grid" Value="True">
            <Setter Property="BackgroundColor" Value="{DynamicResource BorderColor}"/>
        </DataTrigger>
    </Grid.Triggers>

```

I hope you or anyone else can help me out on this!

Thanks in advance and kind regards,

Jeffrey

**Comment by gaurakshay** [comment, marked as answer]
This is how I implemented `TouchEffect` in my View with binding commands to my associated ViewModel.

```
<CollectionView
    Grid.Row="2"
    ItemsSource="{Binding <CollectionSource>}"
    SelectionMode="None">
    <CollectionView.ItemTemplate>
        <DataTemplate x:DataType="<namespace:DataType>">
            <Grid
                xct:TouchEffect.Command="{Binding Source={RelativeSource AncestorType={x:Type <namespace:ViewModel>}}, Path=<CommandName>}"
                xct:TouchEffect.CommandParameter="{Binding Source={RelativeSource Self}, Path=BindingContext}"
                RowDefinitions="Auto">
                <Label
                    Text="{Binding <Property>}"
                    TextColor="Black" />
            </Grid>
        </DataTemplate>
    </CollectionView.ItemTemplate>
</CollectionView>
```

Hope this helps!

**Reply by klingu** [reply]
@gaurakshay Thanks for your example and your quick respons!

When I implemented your line of code
`xct:TouchEffect.Command="{Binding Source={RelativeSource AncestorType={x:Type <TheFullNamespace:MessagesListViewModel>}}, Path=LongPressCommand}"
`
I endup with the error: Error XFC0000: Cannot resolve type "MessagesListViewModel". (XFC0000) (<ProjectName>) XamlC

After some investigation I endup with this solution based ont your implementation:

`xct:TouchEffect.LongPressCommand="{Binding Source={RelativeSource Mode=FindAncestorBindingContext, AncestorType={x:Type messagePages:MessagesListViewModel}}, Path=LongPressCommand}"`

It works!

Maybe the error is Xamarin version related, as I googled it, I found this issue back over the years by upgrading to newer versions of Xamarin.

Thanks for your quick, clear and good support!

Jeffrey

**Comment by AndreiMisiukevich** [comment]
Hey Jeffrey, thanks for asking.
Can you try what Akshay proposed?

If it doesn't help, could you please build a small sample demonstrating the problem?
Because I don't see in your xaml that you apply TouchEffect to any view inside DataTemplate
